src/librbd/librbd.cc

```diff
@@ -4514,10 +4514,12 @@ extern "C" int rbd_group_list(rados_ioctx_t p, char *names, size_t *size)
     return -ERANGE;
   }
 
-  if (!names)
+  if (names == NULL) {
+    tracepoint(librbd, group_list_exit, -EINVAL);
     return -EINVAL;
+  }
 
-  names[expected_size] = '\0';
+  names[expected_size - 1] = '\0';
```

Comment:
If there are no groups, this will fail:

> 2018-02-26T20:01:29.490 INFO:tasks.workunit.client.0.smithi132.stderr:test_rbd.test_list_groups_empty ... ==7424== Invalid write of size 1
> 2018-02-26T20:01:29.490 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    at 0x20B70F53: rbd_group_list (librbd.cc:4533)
> 2018-02-26T20:01:29.490 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x208C7D05: __pyx_pf_3rbd_3RBD_44group_list (rbd.c:10889)
> 2018-02-26T20:01:29.490 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x208C7D05: __pyx_pw_3rbd_3RBD_45group_list (rbd.c:10773)
> 2018-02-26T20:01:29.490 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4E819A2: PyObject_Call (in /usr/lib64/libpython2.7.so.1.0)
> 2018-02-26T20:01:29.490 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4F160F5: PyEval_EvalFrameEx (in /usr/lib64/libpython2.7.so.1.0)
> 2018-02-26T20:01:29.490 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4F1CEFC: PyEval_EvalCodeEx (in /usr/lib64/libpython2.7.so.1.0)
> 2018-02-26T20:01:29.490 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4EA6857: ??? (in /usr/lib64/libpython2.7.so.1.0)
> 2018-02-26T20:01:29.491 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4E819A2: PyObject_Call (in /usr/lib64/libpython2.7.so.1.0)
> 2018-02-26T20:01:29.491 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4F155BC: PyEval_EvalFrameEx (in /usr/lib64/libpython2.7.so.1.0)
> 2018-02-26T20:01:29.491 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4F1A57C: PyEval_EvalFrameEx (in /usr/lib64/libpython2.7.so.1.0)
> 2018-02-26T20:01:29.491 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4F1CEFC: PyEval_EvalCodeEx (in /usr/lib64/libpython2.7.so.1.0)
> 2018-02-26T20:01:29.491 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4EA694C: ??? (in /usr/lib64/libpython2.7.so.1.0)
> 2018-02-26T20:01:29.491 INFO:tasks.workunit.client.0.smithi132.stderr:==7424==    by 0x4E819A2: PyObject_Call (in /usr/lib64/libpython2.7.so.1.0)
>

Reply:
@dillaman I am not sure about returning expected size 1 for empty list.

It looks like `rbd_group_list` was made similar to `rbd_list`, and `rbd_list` returns 0 for the empty list. It looks like it is expected that the caller should check the returned value (see e.g. my "empty directory rbd_list test" commit I added to this PR).

I suppose we want `rbd_list` and `rbd_group_list` behave similar, and I suppose it is too late to change `rbd_list` behavior? So I just deleted `names[expected_size] = '\0'` which looks like unnecessary here (and `rbd_list` does not have it).

Additionaly, I added the fix for similar tracepoint leak in `rbd_list`, and also unified `rbd_group_image_list` and `rbd_group_snap_list` behavior.